File: Binary Tree/java/src/BinarySearchTrees.java
import java.util.LinkedList;
import java.util.Queue;

public class BinarySearchTrees {
    Node<Integer> root;

    BinarySearchTrees() {
        root = null;
    }

    private Node<Integer> addRecursive(Node<Integer> current, int value) {

        if (current == null) {

            return new Node<Integer>(value);
        }

        if (current.value > value) {
            current.left = addRecursive(current.left, value);
        } else if (value > current.value) {
            current.right = addRecursive(current.right, value);
        }

        return current;

    }

    void add(int value) {
        root = addRecursive(root, value);
    }

    private boolean containsNodeRecursive(Node<Integer> current, int value) {
        if (current == null) {
            return false;
        }
        if (value == current.value) {
            return true;
        }
        return value < current.value
                ? containsNodeRecursive(current.left, value)
                : containsNodeRecursive(current.right, value);
    }

    Boolean search(int value) {

        return containsNodeRecursive(root, value);
    }

    void inOrder() {
        new DepthFirstSearch<Integer>(root).displayInOrder();
    }

    void preOrder() {
        new DepthFirstSearch<Integer>(root).displayPreOrder();
    }

    void postOrder() {
        new DepthFirstSearch<Integer>(root).displayPostOrder();
    }

    void breadthFirstSearch() {
        if (root == null) {
            return;
        }
        Queue<Node<Integer>> nodes = new LinkedList<>();

        Node<Integer> temp = root;

        nodes.add(temp);

        System.out.print("[");
        while (!nodes.isEmpty()) {
            temp = nodes.remove();

            System.out.print(temp.value + ",");

            if (temp.left != null) {
                nodes.add(temp.left);
            }

            if (temp.right != null) {
                nodes.add(temp.right);
            }
        }
        System.out.println("]");
    }

    public void remove(int value) {
        root = deleteRecursive(root, value);
    }

    private Node<Integer> deleteRecursive(Node<Integer> current, int value) {
        if (current == null) {
            return null;
        }

        if (value == current.value) {

            current = deleteNode(current);
            return current;
        }
        if (value < current.value) {
            current.left = deleteRecursive(current.left, value);
            return current;
        }
        current.right = deleteRecursive(current.right, value);
        return current;
    }

    private Node<Integer> deleteNode(Node<Integer> current) {
        // a node has no children
        if (current.left == null && current.right == null) {
            return null;
        }
        // a node has exactly one child
        if (current.right == null) {
            return current.left;
        }

        if (current.left == null) {
            return current.right;
        }

        // a node has two children
        int smallestValue = findSmallestValue(current.right);
        current.value = smallestValue;
        current.right = deleteRecursive(current.right, smallestValue);
        return current;

    }

    private int findSmallestValue(Node<Integer> root) {
        return root.left == null ? root.value : findSmallestValue(root.left);
    }

}

File: Binary Tree/python/binarySearchTrees.py

from node import Node
from queue import Queue
from depthFirstSearch import DepthFirstSearch


class BinarySearchTrees:
    def __init__(self) -> None:
        self.root: Node = None

    def __insert(self, root: Node, value: int) -> Node:

        if (root == None):
            return Node(value=value)
        if (root.value > value):
            root.left = self.__insert(root.left, value)
        elif (value > root.value):
            root.right = self.__insert(root.right, value)

        return root

    def insert(self, value: int) -> None:
        self.root = self.__insert(self.root, value)

    def inOrder(self) -> None:
        DepthFirstSearch(self.root).displayInOrder()

    def preOrder(self) -> None:
        DepthFirstSearch(self.root).displayPreOrder()

    def postOrder(self) -> None:
        DepthFirstSearch(self.root).displayPostOrder()

    def search(self, value: int) -> bool:
        return self.__search(self.root, value)

    def __search(self, root: Node, value: int) -> bool:
        if (root == None):
            return False
        if (root.value == value):
            return True

        elif (root.value > value):
            return self.__search(root.left, value)

        else:
            return self.__search(root.right, value)

    def remove(self, value: int) -> None:
        self.root = self.__remove(self.root, value)

    def __remove(self, root: Node, value: int) -> Node:
        if (root == None):
            return None

        if (root.value == value):
            root = self.__deleteNode(root)

            return root
        elif (root.value > value):
            root.left = self.__remove(root.left, value)
            return root
        else:
            root.right = self.__remove(root.right, value)
            return root

    def __deleteNode(self, root: Node) -> Node:
        #  a node has no children
        if (root.left == None and root.right == None):

            return None

        #  a node has exactly one child
        elif (root.right == None):
            return root.left
        elif (root.left == None):
            return root.right

        #  a node has two children
        smallestValue = self.__findSmallestVale(root.right)
        root.value = smallestValue
        root.right = self.__remove(root.right, smallestValue)
        return root

    def __findSmallestVale(self, root: Node) -> int:
        return root.value if (root.left == None) else self.__findSmallestVale(root.left)

    def breadthFirstSearch(self)-> None:
        if(self.root == None):
            return
        
        q:Queue = Queue()
        
        temp:Node = self.root
        
        q.put(temp)
        
        print("[", end="")
        while(not q.empty()):
            temp = q.get()
            
            print(temp.value, end=",")
            
            if (temp.left != None):
                q.put(temp.left)
            
            if(temp.right != None):
                q.put(temp.right)
        print("]")